src/utils/proxyMap.ts

```diff
@@ -0,0 +1,121 @@
+import { proxy } from 'valtio'
+
+type KeyValRecord<K, V> = {
+  key: K
+  value: V
+}
+
+type InternalProxyMap<K, V> = Map<K, V> & {
+  data: KeyValRecord<K, V>[]
+  toJSON: object
+}
+
+/**
+ * proxyMap
+ *
+ * This is to create a proxy which mimic the native Map behavior.
+ * The API is the same as Map API
+ *
+ * @example
+ * import { proxyMap } from 'valtio/utils'
+ * const state = proxyMap([["key", "value"]])
+ *
+ * //can be used inside a proxy as well
+ * const state = proxy({
+ *   count: 1,
+ *   map: proxyMap()
+ * })
+ *
+ * // When using an object as a key, you can wrap it with `ref` so it's not proxied
+ * // this is useful if you want to preserve the key equality
+ * import { ref } from 'valtio'
+ *
+ * const key = ref({})
+ * state.set(key, "value")
+ * state.get(key) //value
+ *
+ * const key = {}
+ * state.set(key, "value")
+ * state.get(key) //undefined
+ */
+export const proxyMap = <K, V>(
+  entries?: Iterable<readonly [K, V]> | null
+): Map<K, V> => {
+  const map = proxy<InternalProxyMap<K, V>>({
+    data: Array.from(entries || [], (v) => ({ key: v[0], value: v[1] })),
+    has(key) {
+      return this.data.some((p) => p.key === key)
+    },
+    set(key, value) {
+      const record = this.data.find((p) => p.key === key)
+      if (record) {
+        record.value = value
+      } else {
+        this.data.push({
+          key,
+          value,
+        })
+      }
+      return this
+    },
+    get(key) {
+      return this.data.find((p) => p.key === key)?.value
+    },
+    delete(key) {
+      const index = this.data.findIndex((p) => p.key === key)
+      if (index === -1) {
+        return false
+      }
+      this.data.splice(index, 1)
+      return true
+    },
+    clear() {
+      this.data.splice(0)
+    },
+    get size() {
+      return this.data.length
+    },
+    toJSON() {
+      return {}
+    },
+    forEach(cb) {
+      this.data.forEach(({ key, value }) => {
+        cb(value, key, this)
+      })
+    },
+    keys() {
+      return this.data.map(({ key }) => key).values()
+    },
+    values() {
+      return this.data.map(({ value }) => value).values()
+    },
+    entries() {
+      const map = new Map<K, V>()
+      this.data.forEach(({ key, value }) => {
+        map.set(key, value)
+      })
+      return map.entries()
```

Comment:
If we were to have `type KeyValueRecord<K, V> = [key: K, value: V]`, this could become:
```ts
    return new Map<K, V>(this.data).entries()
```

Just an idea, not a suggestion.

Reply:
I like your idea. Especially the fact that we can create the map in one step. I'll switch the implementation if you are ok with it

Reply:
Using an array also simplify the initialization of data
```ts
 data: Array.from(entries || [], (v) => ({ key: v[0], value: v[1] })),


 data: Array.from(entries || []) as KeyValRecord<K, V>[] // the assertion is needed as entries is readonly
```

Reply:
I guess you need to copy the entry items? hm, but `proxy()` will copy it anyway. so, yeah, type assertion is actually fine.